### PR TITLE
Update OpenSSL to version 3.0.9 in the base system

### PR DESCRIPTION
**Pull request by khorben**
This branch is the closest I am currently to a functional update to OpenSSL 3.0.9 in FreeBSD's base system. It was started from an update to the `vendor/openssl-3.0` branch. (see https://reviews.freebsd.org/D40365)

The thorough review expected should include:
* The configuration applied (default from the corresponding `security/openssl30` port with the legacy provider enabled)
* Security patches up to date (the 3.0.9 version was released on May 30th, 2023)
* Other patches in the corresponding port may be relevant too (I have not included them)
* API changes reflected accurately in `secure/lib/libcrypto/Version.map` and `secure/lib/libssl/Version.map`
* Whether the choice of `30` as `SHLIB_MAJOR` is good for the `.so` files (upstream's `3` has already been obsoleted in FreeBSD's base system)
* Functional tests on Tier-1 architectures (amd64, aarch64...)
* Relevant and functional `ossl-modules` providers
* Patches to software using OpenSSL are correct (currently libarchive, dumpon, and Kerberos)
* This does not touch the assembly OpenSSL files that were moved into `sys/crypto/openssl` to avoid any trouble with the kernel, and re-imports them into `secure/lib/libcrypto/arch` instead as per `Makefile.asm`; security fixes beware of both locations.

In most software users of OpenSSL, a compatibility compilation flag was used in order to expose and use the former OpenSSL 1.1 API, which is still provided by OpenSSL 3.0 on request.

PR: 271615
Sponsored by: The FreeBSD Foundation

### CODE REVIEW

**Comment by khorben** [comment]
cc @bsdjhb @emaste

**Comment by khorben** [comment]
I realized moments ago that I haven't taken care of the manual pages yet; this will also require some work here.

**Comment by khorben** [comment]
The new manual pages are now generated, and the obsolete files listed.

**Comment by khorben** [comment]
`9669855118afb40b1f8ec33ae34e4030b427f9dc` is the HEAD of my candidate for inclusion.
The two checks failing here do not seem to be related to these changes to me; they both choke like this:
```
===> usr.bin/factor/tests (install)
install -N /tmp/cirrus-ci-build/etc -U -M /usr/obj/tmp/cirrus-ci-build/amd64.amd64/worldstage//METALOG -D /usr/obj/tmp/cirrus-ci-build/amd64.amd64/worldstage -T package=tests -o root  -g wheel -m 555  factor_tests  /usr/obj/tmp/cirrus-ci-build/amd64.amd64/worldstage/usr/tests/usr.bin/factor/factor_tests
install: /usr/obj/tmp/cirrus-ci-build/amd64.amd64/worldstage/usr/tests/usr.bin/factor/factor_tests: No such file or directory
*** Error code 71
```

**Comment by kevans91** [comment]
> [9669855](https://github.com/freebsd/freebsd-src/commit/9669855118afb40b1f8ec33ae34e4030b427f9dc) is the HEAD of my candidate for inclusion. The two checks failing here do not seem to be related to these changes to me; they both choke like this:
> 
> ```
> ===> usr.bin/factor/tests (install)
> install -N /tmp/cirrus-ci-build/etc -U -M /usr/obj/tmp/cirrus-ci-build/amd64.amd64/worldstage//METALOG -D /usr/obj/tmp/cirrus-ci-build/amd64.amd64/worldstage -T package=tests -o root  -g wheel -m 555  factor_tests  /usr/obj/tmp/cirrus-ci-build/amd64.amd64/worldstage/usr/tests/usr.bin/factor/factor_tests
> install: /usr/obj/tmp/cirrus-ci-build/amd64.amd64/worldstage/usr/tests/usr.bin/factor/factor_tests: No such file or directory
> *** Error code 71
> ```

Note that 8ec966976333d16ec28d4225170c5b957d14cc76 reverts the pertinent part of `etc/mtree/BSD.tests.dist` that would've prevented that particular error.

**Comment by jurajlutter** [comment]
@khorben please, could you rebase on top of today's HEAD?

**Comment by khorben** [comment]
> @khorben please, could you rebase on top of today's HEAD?

Hi @jurajlutter, sorry I got sick yesterday; I have just managed to push a first attempt at rebasing at https://github.com/khorben/freebsd-src/tree/khorben/openssl-3.0.9-rebased.
I haven't build-tested it yet.

**Comment by khorben** [comment]
> Note that [8ec9669](https://github.com/freebsd/freebsd-src/commit/8ec966976333d16ec28d4225170c5b957d14cc76) reverts the pertinent part of `etc/mtree/BSD.tests.dist` that would've prevented that particular error.

I had fixed this issue up the chain of commits in 435a2ee4dd8f5c0b522a723736f7f1259002b589 already when creating this PR, so this shouldn't be related.

**Comment by kevans91** [comment]
> > Note that [8ec9669](https://github.com/freebsd/freebsd-src/commit/8ec966976333d16ec28d4225170c5b957d14cc76) reverts the pertinent part of `etc/mtree/BSD.tests.dist` that would've prevented that particular error.
> 
> I had fixed this issue up the chain of commits in [435a2ee](https://github.com/freebsd/freebsd-src/commit/435a2ee4dd8f5c0b522a723736f7f1259002b589) already when creating this PR, so this shouldn't be related.

It's not clear that we're talking about the same thing. In the current version of this branch, `factor` is still missing from etc/mtree/BSD.tests.dist: https://github.com/khorben/freebsd-src/blob/khorben/openssl-3.0.9/etc/mtree/BSD.tests.dist#L1039 

There's no commit after 8ec966976333d16ec28d4225170c5b957d14cc76 in this branch that puts `factor` back.

**Comment by khorben** [comment]
> It's not clear that we're talking about the same thing. In the current version of this branch, `factor` is still missing from etc/mtree/BSD.tests.dist: https://github.com/khorben/freebsd-src/blob/khorben/openssl-3.0.9/etc/mtree/BSD.tests.dist#L1039
> 
> There's no commit after [8ec9669](https://github.com/freebsd/freebsd-src/commit/8ec966976333d16ec28d4225170c5b957d14cc76) in this branch that puts `factor` back.

Yes I'm sorry I remembered an issue with `ObsoleteFiles.inc` and looked at that instead. I believe the latest push at 5c1525710a1ad6b3449acefbcf62292355e3116b actually addresses what you reported; thanks!

**Comment by kevans91** [comment]
> > It's not clear that we're talking about the same thing. In the current version of this branch, `factor` is still missing from etc/mtree/BSD.tests.dist: https://github.com/khorben/freebsd-src/blob/khorben/openssl-3.0.9/etc/mtree/BSD.tests.dist#L1039
> > There's no commit after [8ec9669](https://github.com/freebsd/freebsd-src/commit/8ec966976333d16ec28d4225170c5b957d14cc76) in this branch that puts `factor` back.
> 
> Yes I'm sorry I remembered an issue with `ObsoleteFiles.inc` and looked at that instead. I believe the latest push at [5c15257](https://github.com/freebsd/freebsd-src/commit/5c1525710a1ad6b3449acefbcf62292355e3116b) actually addresses what you reported; thanks!

Yup, thanks! I had opened PR khorben/freebsd-src#2 in case that was easier, but I've closed that now as it's been addressed and install looks happier now.

**Comment by vstakhov** [comment]
It seems you have forgotten to add file `ec/ec_deprecated.c` to `SRCS` in `secure/lib/libcrypto/Makefile`. It causes troubles with some symbols that are defined in the includes but missing from the `libcrypto`.

File: secure/lib/libcrypto/Makefile [hunk 1]
```diff
@@ -210,8 +210,8 @@ SRCS+=	dso_dlfcn.c dso_err.c dso_lib.c
 # ec
 SRCS+=	curve25519.c curve448.c curve448_tables.c ec2_oct.c ec2_smpl.c
 SRCS+=	ec_ameth.c ec_asn1.c ec_backend.c ec_check.c ec_curve.c ec_cvt.c
-SRCS+=	ec_err.c ec_key.c ec_kmeth.c ec_lib.c ec_mult.c ec_oct.c ec_pmeth.c
-SRCS+=	ec_print.c ecdh_kdf.c ecdh_ossl.c ecdsa_ossl.c ecdsa_sign.c
+SRCS+=	ec_deprecated.c ec_err.c ec_key.c ec_kmeth.c ec_lib.c ec_mult.c ec_oct.c
```
Comment (vstakhov) [inline secure/lib/libcrypto/Makefile:213]: 👍

**Comment by jurajlutter** [comment]
For me: ```In file included from /usr/src/lib/libradius/radlib.c:38:                                                                                              In file included from /usr/obj/usr/src/amd64.amd64/tmp/usr/include/openssl/hmac.h:14:                                                                  /usr/obj/usr/src/amd64.amd64/tmp/usr/include/openssl/macros.h:139:4: error: "The requested API level higher than the configured API compatibility level"                                                                                                                                                      #  error "The requested API level higher than the configured API compatibility level"```

**Comment by khorben** [comment]
Hi Juraj,

> For me:
> `In file included from /usr/src/lib/libradius/radlib.c:38: In file included from`
> `/usr/obj/usr/src/amd64.amd64/tmp/usr/include/openssl/hmac.h:14:`
> `/usr/obj/usr/src/amd64.amd64/tmp/usr/include/openssl/macros.h:139:4:`
> `error: "The requested API level higher than the configured API compatibility level"`
> `# error "The requested API level higher than the configured API compatibility level"`

Would you have more information about your build environment? Ideally the contents of your `make.conf` file if you configured it, or any other build option?

**Comment by emaste** [comment]
b077aed33b7b6aefca7b17ddb250cf521f938613

**Pull request closed without merging**